core/arch/riscv64/memfuncs.asm

```diff
@@ -44,30 +44,130 @@ START_FILE
  */
         DECLARE_FUNC(memcpy)
 GLOBAL_LABEL(memcpy:)
-/* TODO i#3544: Naive version, optimize it. */
-        mv       t1, ARG1
-        beqz     ARG3, 2f
-1:      lbu      t2, 0(ARG2)
-        addi     ARG2, ARG2, 1
-        sb       t2, 0(t1)
-        addi     t1, t1, 1
-        addi     ARG3, ARG3, -1
-        bnez     ARG3, 1b
-2:      ret
+        addi     a5, x0, 32
+        mv       t0, ARG2 /* save for return */
+copy32_:
+        blt      ARG3, a5, copy0_32
+        ld       t1, 0(ARG2)
+        ld       t2, 8(ARG2)
+        ld       t3, 16(ARG2)
+        ld       t4, 24(ARG2)
+        sd       t1, 0(ARG1)
+        sd       t2, 8(ARG1)
+        sd       t3, 16(ARG1)
+        sd       t4, 24(ARG1)
+        addi     ARG3, ARG3, -32
+        addi     ARG1, ARG1, 32
+        addi     ARG2, ARG2, 32
+        j        copy32_
+copy0_32:
+        add      a6, ARG2, ARG3 /* a6 = src_end */
+        add      a7, ARG1, ARG3 /* a7 = dst_end */
+        addi     a5, x0, 8
+        bge      ARG3, a5, copy8_32
+        addi     a5, x0, 4
+        bge      ARG3, a5, copy4_8
+        bgtz     ARG3, copy0_4
+        j        copyexit
+copy0_4:
+        srli     t4, ARG3, 1
+        add      t5, t4, ARG1
+        add      t4, t4, ARG2
+        lbu      t1, 0(ARG2)
+        lbu      t2, -1(a6)
+        lbu      t3, 0(t4)
```

Comment:
Again, please add comments as this is not simple to understand.  I *think* this is using tricks to avoid branches?  It's using tricks to get the 2nd and 3rd byte to be redundant if the size is just 1 or 2?  This type of thing really needs a comment to make the code readable and maintainable.

Reply:
Right. Exactly a trick here.